lib/Dialects/LinalgTransform/Transforms/TrackingRewriteDriver.cpp

```diff
@@ -0,0 +1,439 @@
+    // patterns that identify the "main" result of a transformation.
+    while (auto castOp = value.getDefiningOp<tensor::CastOp>())
+      value = castOp.source();
+
+    if (auto currentSourceOp = value.getDefiningOp<LinalgOp>()) {
+      if (!sourceOp || sourceOp == currentSourceOp) {
+        sourceOp = currentSourceOp;
+        continue;
+      }
+
+      LLVM_DEBUG(
+          DBGS() << "different source linalg ops for replacing one op: \n"
+                 << sourceOp << "\n"
+                 << currentSourceOp << "\n");
+      return nullptr;
+    }
+  }
+  return sourceOp;
+}
+
+// TODO: if GreedyPatternRewriteDriver had been public, we would have derived
+// it, but we have to copy for now.
+class TrackingRewriteDriver : public PatternRewriter {
+public:
+  explicit TrackingRewriteDriver(
+      MLIRContext *ctx, DenseMap<Value, Operation *> &trackedOperations,
+      const FrozenRewritePatternSet &patterns,
+      const GreedyRewriteConfig &config)
+      : PatternRewriter(ctx), trackedOperations(trackedOperations),
+        matcher(patterns), folder(ctx), config(config) {
+    worklist.reserve(64);
+    matcher.applyDefaultCostModel();
+
+    for (auto &pair : trackedOperations)
+      trackedOperationKeys.try_emplace(pair.second, pair.first);
+  }
+
+  void replaceOp(Operation *op, ValueRange newValues) override {
+    if (isa<LinalgOp>(op)) {
+      LinalgOp replacement = findSingleSourceOp(newValues);
+      assert(replacement &&
+             "repacling a linalg op, but could not find the replacement op");
+
+      if (Value key = trackedOperationKeys.lookup(op)) {
+        LLVM_DEBUG(DBGS() << "replacing tracked " << *op << " with "
+                          << *replacement << "for " << key << "\n");
+        trackedOperations[key] = replacement;
+        trackedOperationKeys.erase(op);
+        trackedOperationKeys.try_emplace(replacement, key);
+      }
+    }
+    PatternRewriter::replaceOp(op, newValues);
+  }
+
+  /// PatternRewriter hook for erasing a dead operation.
+  void eraseOp(Operation *op) override {
+    LLVM_DEBUG({
+      logger.startLine() << "** Erase   : '" << op->getName() << "'(" << op
+                         << ")\n";
+    });
+    assert(trackedOperationKeys.count(op) == 0 &&
+           "attempting to erase a tracked op");
+    PatternRewriter::eraseOp(op);
+  }
+
+  // TODO: if notifyRootReplaced had been also getting the ValueRange, we
+  // wouldn't need to override these.
+  void
+  replaceOpWithIf(Operation *, ValueRange, bool *,
+                  llvm::unique_function<bool(OpOperand &) const>) override {
+    llvm_unreachable("TrackingRewriteDriver::replaceOpWithIf not implemented");
+  }
+
+private:
+  DenseMap<Value, Operation *> &trackedOperations;
+  DenseMap<Operation *, Value> trackedOperationKeys;
+
+  //--------------------------------------------------------------------------//
+  // The code below is taken from GreedyPatternRewriteDriver.
```

Comment:
Ok, so I am skipping reviewing until the matching `end` without risk of errors.
Please shout if incorrect.